allennlp/data/iterators/adaptive_iterator.py

```diff
@@ -39,67 +60,43 @@ class AdaptiveIterator(BucketIterator):
-        (1) Provide the ``padding_memory_scaling`` function, which gives a big-O bound on memory
-        usage given padding lengths. For instance, if you have two TextFields with ``sentence_lengths``
-        which require padding, this might be simply |sentence1| * |sentence2|.
-        (2) Tune the `adaptive_memory_usage_constant` parameter for your particular model and GPU.
-        The only reason this is a parameter is so that you can do this tuning easily by setting this to
-        False. If you are not using adaptive batch sizes, you should use a :class:`~BucketIterator`.
-    biggest_batch_first: bool, optional (default=False)
-        This is largely for testing, to see how large of a batch you can safely use with your GPU.
-        It's only meaningful if you're using dynamic padding - this will let you try out the
-        largest batch that you have in the data `first`, so that if you're going to run out of
-        memory, you know it early, instead of waiting through the whole batch to find out at the
-        end that you're going to crash.
-
-    See :class:`~BucketIterator` for a description of the other parameters.
+    biggest_batch_first : bool, optional (default=False)
+        See :class:`BucketIterator`.  If this is ``True``, we bypass the adaptive grouping step, so
+        you can tune the ``adaptive_memory_usage_constant``.
+    sorting_keys : List[Tuple[str, str]]
+        See :class:`BucketIterator`.
+    padding_noise : List[Tuple[str, str]]
+        See :class:`BucketIterator`.
     """
     def __init__(self,
                  adaptive_memory_usage_constant: float,
-                 padding_memory_scaling: Callable[Dict[str, int], float],
+                 padding_memory_scaling: Callable[[Dict[str, int]], float],
                  maximum_batch_size: int = 10000,
-                 use_adaptive_grouping: bool = True,
                  biggest_batch_first: bool = False,
                  sorting_keys: List[Tuple[str, str]] = None,
-                 padding_noise: float = 0.2,
-                 sort_every_epoch: bool = True):
-
-        self.padding_memory_scaling = padding_memory_scaling
-        self.maximum_batch_size = maximum_batch_size
-        self.adaptive_memory_usage_constant = adaptive_memory_usage_constant
-        self.biggest_batch_first = biggest_batch_first
-        self.use_adaptive_grouping = use_adaptive_grouping
-        super(AdaptiveIterator, self).__init__(sorting_keys, padding_noise, sort_every_epoch)
+                 padding_noise: float = 0.2):
+        self._padding_memory_scaling = padding_memory_scaling
+        self._maximum_batch_size = maximum_batch_size
+        self._adaptive_memory_usage_constant = adaptive_memory_usage_constant
+        super(AdaptiveIterator, self).__init__(sorting_keys, padding_noise, biggest_batch_first)
 
     @overrides
-    def _create_batches(self, dataset: Dataset) -> List[List[Instance]]:
+    def num_batches_per_epoch(self, dataset: Dataset) -> int:
```

Comment:
Does this concept even make sense for an adaptive iterator? If padding noise is non-zero, this is non-deterministic which is a little counter intuitive (and makes the argument for not having this as a side effect of `__call__` weaker) and having to actually process the dataset just to return this value seems a little heavy.